src/bluesearch/database/topic_info.py

```diff
@@ -0,0 +1,116 @@
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+"""Implementation of the TopicInfo data structure."""
+from __future__ import annotations
+
+import copy
+import datetime
+import pathlib
+from dataclasses import dataclass, field
+from typing import Any
+
+import bluesearch
+from bluesearch.database.article import ArticleSource
+
+
+@dataclass
+class TopicInfo:
+    """The topic information extracted from a journal article.
+
+    For the spec see the following GitHub issue/comment:
+    https://github.com/BlueBrain/Search/issues/518#issuecomment-985525160
+    """
+
+    source: ArticleSource
+    path: str | pathlib.Path
+    element_in_file: int | None = None
+    article_topics: dict[str, list[str]] = field(init=False, default_factory=dict)
+    journal_topics: dict[str, list[str]] = field(init=False, default_factory=dict)
+    metadata: dict = field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        """Run the post-initialization."""
+        self.creation_date = datetime.datetime.now()
+
+    @staticmethod
+    def add_topics(mapping: dict[str, list[str]], kind: str, topics: list[str]) -> None:
+        """Add topics to a mapping with collection of topics.
+
+        Parameters
+        ----------
+        mapping
+            A mapping of the form kind -> list-of-topics that shall be
+            updated in-place. For example ``{"MeSH": ["topic 1", "topic 2"]}``.
+        kind
+            The topic kind. Corresponds to a key in ``mapping``.
+        topics
+            The topics to add. Corresponds to a value in ``mapping``.
+        """
+        updated_topics = mapping.get(kind, []) + topics
+        mapping[kind] = sorted(set(updated_topics))
+
+    def add_article_topics(self, kind: str, topics: list[str]) -> None:
+        """Add article topics.
+
+        Parameters
+        ----------
+        kind
+            The topic kind. For example "MeSH" or "MAG".
+        topics
+            A list of the topics to add.
+        """
+        self.add_topics(self.article_topics, kind, topics)
+
+    def add_journal_topics(self, kind: str, topics: list[str]) -> None:
+        """Add journal topics.
+
+        Parameters
+        ----------
+        kind
+            The topic kind. For example "MeSH" or "MAG".
+        topics
+            A list of the topics to add.
+        """
+        self.add_topics(self.journal_topics, kind, topics)
+
+    def json(self) -> dict:
+        """Convert the contents of this class to a structured dictionary.
+
+        Returns
+        -------
+        dict
+            The structure dictionary with all topic information.
+        """
+        metadata: dict[str, Any] = {
+            **copy.deepcopy(self.metadata),
+            "created-date": self.creation_date.strftime("%Y-%m-%d %H:%M:%S"),
+            "bbs-version": bluesearch.__version__,
+        }
+        if self.element_in_file is not None:
+            metadata["element_in_file"] = self.element_in_file
```

Comment:
Just for me to understand, what do we expect inside `self.metadata` ? 
Shouldn't `element_in_file` be part of the metadata  ?

Reply:
Right now it's just possible to pass any additional metadata via `TopicInfo(..., metadata={"some key": "some value", ...})`, that's what will be in `self.metadata`. But if we don't want that, I guess we can get rid of that. I'm actually not so sure here about what we want exactly.

I'm not sure I understand the question about `element_in_file`? It's inserted into `metadata` if it's set, otherwise it will be absent. At least that's how it was specified in https://github.com/BlueBrain/Search/issues/518#issuecomment-985525160.

Reply:
Sorry, I am realising that my comment was not really clear and maybe not at the right place. I was wondering if `element_in_file` should be a separate attribute or if we should expect the user to give this info through the metadata dictionary (as it is really source specific and going to end in the metadata field anyway) but maybe it is better to keep it explicit.

Reply:
Ah I see :) There's actually a discussion thread with @FrancescoCasalegno about the `metadata` parameter further below. For now I completely removed it in 827c79d. So I guess for now it has to stay as a separate paramter.

Reply:
But feel free to chime in on the structure discussion below, I think we're still looking for a good solution :)